--- LVM2.2.02.72-RHEL6-orig/lib/activate/activate.c	2010-07-12 13:38:35.000000000 +0100
+++ LVM2.2.02.72-RHEL6-edited/lib/activate/activate.c	2010-08-18 01:58:34.000000000 +0100
@@ -762,18 +762,14 @@ int monitor_dev_for_events(struct cmd_co
 
 	/*
 	 * In case this LV is a snapshot origin, we instead monitor
-	 * each of its respective snapshots (the origin itself does
-	 * not need to be monitored).
-	 *
-	 * TODO: This may change when snapshots of mirrors are allowed.
+	 * each of its respective snapshots.  The origin itself may
+	 * also need to be monitored if it is a mirror, for example.
 	 */
-	if (lv_is_origin(lv)) {
+	if (lv_is_origin(lv))
 		dm_list_iterate_safe(snh, snht, &lv->snapshot_segs)
 			if (!monitor_dev_for_events(cmd, dm_list_struct_base(snh,
 				    struct lv_segment, origin_list)->cow, monitor))
 				r = 0;
-		return r;
-	}
 
 	/*
 	 * If the volume is mirrored and its log is also mirrored, monitor
--- LVM2.2.02.72-RHEL6-orig/lib/mirror/mirrored.c	2010-07-21 16:21:24.000000000 +0100
+++ LVM2.2.02.72-RHEL6-edited/lib/mirror/mirrored.c	2010-08-18 01:57:21.000000000 +0100
@@ -604,7 +604,8 @@ static int _target_monitored(struct lv_s
 	if (!_get_mirror_dso_path(vg->cmd, &dso))
 		return_0;
 
-	if (!(uuid = build_dm_uuid(vg->cmd->mem, lv->lvid.s, NULL)))
+	/* We always monitor the "real" device, never the "snapshot-origin" itself. */
+	if (!(uuid = build_dm_uuid(vg->cmd->mem, lv->lvid.s, lv_is_origin(lv) ? "real" : NULL)))
 		return_0;
 
 	if (!(dmevh = _create_dm_event_handler(uuid, dso, DM_EVENT_ALL_ERRORS)))
@@ -642,7 +643,8 @@ static int _target_set_events(struct lv_
 	if (!_get_mirror_dso_path(vg->cmd, &dso))
 		return_0;
 
-	if (!(uuid = build_dm_uuid(vg->cmd->mem, lv->lvid.s, NULL)))
+	/* We always monitor the "real" device, never the "snapshot-origin" itself. */
+	if (!(uuid = build_dm_uuid(vg->cmd->mem, lv->lvid.s, lv_is_origin(lv) ? "real" : NULL)))
 		return_0;
 
 	if (!(dmevh = _create_dm_event_handler(uuid, dso, DM_EVENT_ALL_ERRORS)))
--- LVM2.2.02.72-RHEL6-orig/WHATS_NEW	2010-08-18 01:45:59.000000000 +0100
+++ LVM2.2.02.72-RHEL6-edited/WHATS_NEW	2010-08-18 01:53:31.000000000 +0100
@@ -1,5 +1,6 @@
 Version 2.02.73 -
 ================================
+  Monitor origin -real device below snapshot instead of overlay device.
   Change default alignment of pe_start to 1MB.
   Require --restorefile when using pvcreate --uuid.
   Recognise and give preference to md device partitions (blkext major).
