 WHATS_NEW             |    1 +
 lib/metadata/mirror.c |   13 +------------
 2 files changed, 2 insertions(+), 12 deletions(-)

diff --git a/WHATS_NEW b/WHATS_NEW
index 4472666..2a8c6f2 100644
--- a/WHATS_NEW
+++ b/WHATS_NEW
@@ -1,5 +1,6 @@
 Version 2.02.73 -
 ================================
+  Fix 'lvconvert --splitmirrors' in cluster operation.
   Ignore snapshots when performing mirror recovery beneath an origin.
   Add suspend_lv_origin and resume_lv_origin using LCK_ORIGIN_ONLY.
   Allow internal suspend and resume of origin without its snapshots.
diff --git a/lib/metadata/mirror.c b/lib/metadata/mirror.c
index b7d5e08..3b1f9dd 100644
--- a/lib/metadata/mirror.c
+++ b/lib/metadata/mirror.c
@@ -719,19 +719,8 @@ static int _split_mirror_images(struct logical_volume *lv,
 		return 0;
 	}
 
-	/*
-	 * Suspend the newly split-off LV (balance memlock count
-	 * and prepare for DM automated renaming via resume).
-	 */
-	if (!suspend_lv(lv->vg->cmd, new_lv)) {
-		log_error("Failed to lock newly split LV, %s", new_lv->name);
-		vg_revert(lv->vg);
-		return 0;
-	}
-
 	/* Bring newly split-off LV into existence */
-	log_very_verbose("Creating %s", new_lv->name);
-	if (!resume_lv(lv->vg->cmd, new_lv)) {
+	if (!activate_lv(lv->vg->cmd, new_lv)) {
 		log_error("Failed to activate newly split LV, %s",
 			  new_lv->name);
 		return 0;
