--- LVM2.2.02.72-RHEL6-edited/lib/activate/dev_manager.c	2010-08-18 01:45:58.000000000 +0100
+++ LVM2.2.02.72-RHEL6-edited2/lib/activate/dev_manager.c	2010-08-18 02:04:44.000000000 +0100
@@ -548,6 +548,7 @@ static int _percent(struct dev_manager *
 	return 0;
 }
 
+/* FIXME Merge with the percent function */
 int dev_manager_transient(struct dev_manager *dm, struct logical_volume *lv)
 {
 	int r = 0;
@@ -558,10 +559,11 @@ int dev_manager_transient(struct dev_man
 	char *type = NULL;
 	char *params = NULL;
 	char *dlid = NULL;
+	const char *layer = lv_is_origin(lv) ? "real" : NULL;
 	const struct dm_list *segh = &lv->segments;
 	struct lv_segment *seg = NULL;
 
-	if (!(dlid = build_dm_uuid(dm->mem, lv->lvid.s, NULL)))
+	if (!(dlid = build_dm_uuid(dm->mem, lv->lvid.s, layer)))
 		return_0;
 
 	if (!(dmt = _setup_task(0, dlid, NULL, DM_DEVICE_STATUS, 0, 0)))
--- LVM2.2.02.72-RHEL6-edited/WHATS_NEW	2010-08-18 01:53:31.000000000 +0100
+++ LVM2.2.02.72-RHEL6-edited2/WHATS_NEW	2010-08-18 02:03:36.000000000 +0100
@@ -1,5 +1,6 @@
 Version 2.02.73 -
 ================================
+  Fix dev_manager_transient to access -real device not snapshot-origin.
   Monitor origin -real device below snapshot instead of overlay device.
   Change default alignment of pe_start to 1MB.
   Require --restorefile when using pvcreate --uuid.
